libs/wallet/src/connectors/snap-connector.ts

```diff
@@ -0,0 +1,233 @@
+        [snapId]: params,
+      },
+    });
+  } catch (err) {
+    // NOOP - rejected by user
+  }
+};
+
+/**
+ * Gets the list of all installed snaps.
+ * More information here: https://docs.metamask.io/snaps/reference/rpc-api/#wallet_getsnaps
+ */
+export const getSnaps = async (): Promise<GetSnapsResponse> => {
+  return (await ethereumRequest({
+    method: 'wallet_getSnaps',
+  })) as GetSnapsResponse;
+};
+
+/**
+ * Gets the requested snap by `snapId` and an optional `version`
+ */
+export const getSnap = async (
+  snapId: string,
+  version?: string
+): Promise<Snap | undefined> => {
+  try {
+    const snaps = await getSnaps();
+    return Object.values(snaps).find(
+      (snap) => snap.id === snapId && (!version || snap.version === version)
+    );
+  } catch (e) {
+    return undefined;
+  }
+};
+
+export const invokeSnap = async <T>(
+  snapId: string,
+  request: InvokeSnapRequest
+) => {
+  const req = {
+    method: 'wallet_invokeSnap',
+    params: {
+      snapId,
+      request,
+    },
+  };
+  return await ethereumRequest<T>(req);
+};
+
+export class SnapConnector implements VegaConnector {
+  description = "Connects using Vega Protocol's MetaMask snap";
+  snapId: string | undefined = undefined;
+  nodeAddress: string | undefined = undefined;
+
+  constructor(nodeAddress?: string, snapId = DEFAULT_SNAP_ID) {
+    this.nodeAddress = nodeAddress;
+    this.snapId = snapId;
+  }
+
+  async listKeys() {
+    if (!this.snapId) throw ERR_SNAP_ID_NOT_SET;
+    return await invokeSnap<ListKeysResponse>(this.snapId, {
+      method: 'client.list_keys',
+    });
+  }
+
+  async connect() {
+    const res = await this.listKeys();
+    setConfig({
+      connector: 'snap',
+      token: null, // no token required for snap
+      url: null, // no url required for snap
+    });
+    return res?.keys;
+  }
+
+  async sendTx(pubKey: string, transaction: Transaction) {
+    if (!this.nodeAddress) throw ERR_NODE_ADDRESS_NOT_SET;
+    if (!this.snapId) throw ERR_SNAP_ID_NOT_SET;
+
+    // This step is needed to strip the transaction object from any additional
+    // properties, such as `__proto__`, etc.
+    let txData = null;
+    try {
+      txData = JSON.parse(JSON.stringify(transaction));
+    } catch (err) {
+      throw ERR_TRANSACTION_PARSE;
+    }
+
+    const payload = {
+      method: 'client.send_transaction',
+      params: {
+        sendingMode: 'TYPE_SYNC',
+        transaction: txData,
+        publicKey: pubKey,
+        networkEndpoints: [this.nodeAddress],
+      },
+    };
+
+    const result = await invokeSnap<SendTransactionResponse>(
+      this.snapId,
+      payload
+    );
+
+    if ('error' in result) {
+      const { message, code, data } = result.error;
+      throw new WalletError(
+        message,
+        code,
+        typeof data === 'string' ? data : ''
+      );
+    }
+
+    return {
+      transactionHash: result.transactionHash,
+      signature: result?.transaction?.signature?.value,
```

Comment:
chore (blocking): can we throw `WalletError` if signature is undefined. This would be better than just type casting the response

```
if (!result?.transaction?.signature) {
  throw new WalletError(message, code)
}

return {
  transactionHash: result.transactionHash,
  signature: result.transaction.signature.value,
  receivedAt: result.receivedAt,
  sentAt: result.sentAt,
};

```

Reply:
I've updated this